router-tests/websocket_test.go

```diff
@@ -2069,6 +2070,238 @@ func TestWebSockets(t *testing.T) {
+			require.NoError(t, err)
+			require.Equal(t, "next", res.Type)
+			require.Equal(t, "1", res.ID)
+			require.Contains(t, string(res.Payload), `"data"`)
+			require.Contains(t, string(res.Payload), `"currentTime"`)
+
+			require.NoError(t, conn.Close())
+		})
+	})
+
+	t.Run("operation with mismatched hash is rejected", func(t *testing.T) {
+		t.Parallel()
+
+		testenv.Run(t, &testenv.Config{}, func(t *testing.T, xEnv *testenv.Environment) {
+			conn := xEnv.InitGraphQLWebSocketConnection(nil, nil, []byte(`{"graphql-client-name": "my-client"}`))
+
+			err := testenv.WSWriteJSON(t, conn, testenv.WebSocketMessage{
+				ID:   "1",
+				Type: "subscribe",
+				Payload: []byte(`{
+					"query": "subscription { currentTime { unixTime } }",
+					"extensions": {
+						"persistedQuery": {
+							"version": 1,
+							"sha256Hash": "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
+						}
+					}
+				}`),
+			})
+			require.NoError(t, err)
+
+			var res testenv.WebSocketMessage
+			err = testenv.WSReadJSON(t, conn, &res)
+			require.NoError(t, err)
+			require.Equal(t, "error", res.Type)
+			require.Equal(t, "1", res.ID)
+			require.Contains(t, string(res.Payload), "persistedQuery sha256 hash does not match query body")
+
+			require.NoError(t, conn.Close())
+		})
+	})
+
+	t.Run("cache poisoning is tried but prevented", func(t *testing.T) {
+		t.Parallel()
+
+		testenv.Run(t, &testenv.Config{
+			ApqConfig: config.AutomaticPersistedQueriesConfig{
+				Enabled: true,
+			},
+		}, func(t *testing.T, xEnv *testenv.Environment) {
+			conn := xEnv.InitGraphQLWebSocketConnection(nil, nil, []byte(`{"graphql-client-name": "my-client"}`))
+
+			query1 := "subscription { employeeUpdated(employeeID: 3) { id } }"
+			query2 := "subscription { currentTime { unixTime } }"
+			hashOfQuery2 := "6e94d99132b544a0d7522696a7d35643d56a26c7b8c2e0df29e2b9935636628c"
+
+			err := testenv.WSWriteJSON(t, conn, testenv.WebSocketMessage{
+				ID:   "1",
+				Type: "subscribe",
+				Payload: []byte(fmt.Sprintf(`{
+					"query": "%s",
+					"extensions": {
+						"persistedQuery": {
+							"version": 1,
+							"sha256Hash": "%s"
+						}
+					}
+				}`, query1, hashOfQuery2)),
+			})
+			require.NoError(t, err)
+
+			var res1 testenv.WebSocketMessage
+			err = testenv.WSReadJSON(t, conn, &res1)
+			require.NoError(t, err)
+			require.Equal(t, "error", res1.Type)
+			require.Equal(t, "1", res1.ID)
+			require.Contains(t, string(res1.Payload), "persistedQuery sha256 hash does not match query body")
+
+			// even though we got an error we challenge that and still try
+			// out if our malicious query was cached with hash2.
+			err = testenv.WSWriteJSON(t, conn, testenv.WebSocketMessage{
+				ID:   "2",
+				Type: "subscribe",
+				Payload: []byte(fmt.Sprintf(`{
+					"query": "%s",
+					"extensions": {
+						"persistedQuery": {
+							"version": 1,
+							"sha256Hash": "%s"
+						}
+					}
+				}`, query2, hashOfQuery2)),
+			})
+			require.NoError(t, err)
```

Comment:
This should send _only_ the hash, not both. The goal is to make sure none of the caches were poisoned which is subverted in some places by the query body being present.

Reply:
fixed